### PR TITLE
M prefix added for script addresses

### PR DESCRIPTION
**Pull request by wangxinxi**
Test cases for M prefix addresses are still needed.

### CODE REVIEW

File: src/chainparams.cpp [hunk 1]
```diff
@@ -126,6 +126,7 @@ class CMainParams : public CChainParams {
 
         base58Prefixes[PUBKEY_ADDRESS] = std::vector<unsigned char>(1,48);
         base58Prefixes[SCRIPT_ADDRESS] = std::vector<unsigned char>(1,5);
+        base58Prefixes[SCRIPT_ADDRESS2] = std::vector<unsigned char>(1,50);
```
Comment (shaolinfry) [inline src/chainparams.cpp]: Please add a SCRIPT_ADDRESS2 for testnet also so we can test it out on testnet.

**Comment by shaolinfry** [comment]
Does this the `addmultisigaddress` RPC command work with the new prefix?

**Comment by wangxinxi** [comment]
`./litecoin-cli addmultisigaddress 2 '["LgebDAqipqUm41KQQHwSGHA56XeyzGiEVf", "LRxuPLUGjYivrjibydfKACLjFKzUiEnXgL"]'`

Output:
MAP2uc4aFVwJwoJp3p8yFMs7zy6Pa5e9Zv

**Comment by thrasher-** [comment]
Looks good 👍 Just need an RPC test to test functionality and then its good to merge.

**Comment by shaolinfry** [comment]
Checked the ability to receive and spend from old and new addresses plus existing RPC calls.

Needs some python RPC tests though.

**Comment by shaolinfry** [comment]
two tests are failing,possibly because of default address prefix change

```
p2p-segwit.py:
start_node: litecoind started, waiting for RPC to come up
start_node: RPC succesfully started
start_node: litecoind started, waiting for RPC to come up
start_node: RPC succesfully started
start_node: litecoind started, waiting for RPC to come up
start_node: RPC succesfully started
MiniNode: Connecting to Litecoin Node IP # 127.0.0.1:13880
MiniNode: Connecting to Litecoin Node IP # 127.0.0.1:13880
MiniNode: Connecting to Litecoin Node IP # 127.0.0.1:13881
Starting tests before segwit lock in:
	Verifying NODE_WITNESS service bit
	Testing non-witness transaction
	Testing behavior of unnecessary witnesses
	Testing block relay
	Testing getblocktemplate setting of segwit versionbit (before lockin)
Testing behavior post lockin, pre-activation
	Testing behavior of unnecessary witnesses
	Testing relay of witness transactions
	Testing block relay
	Testing P2SH witness transactions
	Testing standardness of v0 outputs (before activation)
Testing behavior after segwit activation
	Testing P2SH witness transactions
Assertion failed: False != True
Stopping nodes
Not cleaning up dir /tmp/testl7r8qy0g/360
Failed
stderr:
   File "/home/travis/build/litecoin-project/litecoin/qa/rpc-tests/test_framework/test_framework.py", line 143, in main
    self.run_test()
  File "/home/travis/build/litecoin-project/litecoin/build/../qa/rpc-tests/p2p-segwit.py", line 1999, in run_test
    self.test_p2sh_witness(segwit_activated=True)
  File "/home/travis/build/litecoin-project/litecoin/build/../qa/rpc-tests/p2p-segwit.py", line 1506, in test_p2sh_witness
    self.test_node.test_witness_block(block, accepted=True, with_witness=segwit_activated)
  File "/home/travis/build/litecoin-project/litecoin/build/../qa/rpc-tests/p2p-segwit.py", line 159, in test_witness_block
    assert_equal(self.connection.rpc.getbestblockhash() == block.hash, accepted)
  File "/home/travis/build/litecoin-project/litecoin/qa/rpc-tests/test_framework/util.py", line 504, in assert_equal
    raise AssertionError("%s != %s"%(str(thing1),str(thing2)))
Pass: False, Duration: 36 s
...........
nulldummy.py:
Initializing test directory /tmp/testhg6_j9l5/355
start_node: litecoind started, waiting for RPC to come up
start_node: RPC succesfully started
Test 1: NULLDUMMY compliant base transactions should be accepted to mempool and mined before activation [430]
Test 2: Non-NULLDUMMY base multisig transaction should not be accepted to mempool before activation
Test 3: Non-NULLDUMMY base transactions should be accepted in a block before activation [431]
Test 4: Non-NULLDUMMY base multisig transaction is invalid after activation
Test 5: Non-NULLDUMMY P2WSH multisig transaction invalid after activation
Test 6: NULLDUMMY compliant base/witness transactions should be accepted to mempool and in block after activation [432]
Assertion failed: 0212ff1baa2262c4dce8e3c9f395e5285e7c8b961f3f8883e443843dbba664b5 != a55b5f80365728c993b61b7386714b619730b2602290d2e18678972b10bb9ed1
Stopping nodes
Not cleaning up dir /tmp/testhg6_j9l5/355
Failed
stderr:
   File "/home/travis/build/litecoin-project/litecoin/qa/rpc-tests/test_framework/test_framework.py", line 143, in main
    self.run_test()
  File "/home/travis/build/litecoin-project/litecoin/build/../qa/rpc-tests/nulldummy.py", line 102, in run_test
    self.block_submit(self.nodes[0], test6txs, True, True)
  File "/home/travis/build/litecoin-project/litecoin/build/../qa/rpc-tests/nulldummy.py", line 139, in block_submit
    assert_equal(node.getbestblockhash(), block.hash)
  File "/home/travis/build/litecoin-project/litecoin/qa/rpc-tests/test_framework/util.py", line 504, in assert_equal
    raise AssertionError("%s != %s"%(str(thing1),str(thing2)))
Pass: False, Duration: 6 s
```

**Comment by wangxinxi** [comment]
Errors in p2p-segwit.py are due to the use of wrong version bits. The function self.build_next_block() produces blocks with version = 4, which are rejected by litecoind. 

Errors in nulldummy.py are also due to wrong version bits used. 

We should fix them in another pull request.

**Comment by losh11** [comment]
Why another PR? No need, we could always change the PR title and outline all commits and describe what each commit fixes.

**Comment by wangxinxi** [comment]
The two PRs are not related. Isn't it better to separate them?

**Comment by shaolinfry** [comment]
If the test breaks are not related to this PR, then this can be merged. I have confirmed those tests do fail on master.

File: src/base58.cpp [hunk 2]
```diff
@@ -212,12 +212,18 @@ class CBitcoinAddressVisitor : public boost::static_visitor<bool>
 {
 private:
     CBitcoinAddress* addr;
```
Comment (Anita713) [inline src/base58.cpp]: 35JsUDH3xx5sDditwhPE3QVjvoVgAtcgMJ

File: src/base58.cpp [hunk 3]
```diff
@@ -212,12 +212,18 @@ class CBitcoinAddressVisitor : public boost::static_visitor<bool>
...
+    CChainParams::Base58Type script_type_;
```
Comment (Anita713) [inline src/base58.cpp]: 35JsUDH3xx5sDditwhPE3QVjvoVgAtcgMJ

File: src/base58.h [hunk 4]
```diff
@@ -104,8 +104,8 @@ class CBase58Data
 class CBitcoinAddress : public CBase58Data {
```
Comment (Anita713) [inline src/base58.h]: 35JsUDH3xx5sDditwhPE3QVjvoVgAtcgMJ

File: src/base58.h [hunk 5]
```diff
@@ -104,8 +104,8 @@ class CBase58Data
...
 public:
     bool Set(const CKeyID &id);
```
Comment (Anita713) [inline src/base58.h]: evansanita713@gmail.com

File: src/base58.h [hunk 6]
```diff
@@ -104,8 +104,8 @@ class CBase58Data
...
-    bool Set(const CScriptID &id);
```
Comment (Anita713) [inline src/base58.h]: 35JsUDH3xx5sDditwhPE3QVjvoVgAtcgMJ

File: src/base58.h (hunk 4 above)
Comment (Anita713) [reply src/base58.h]: 256

File: src/base58.h [hunk 7]
```diff
@@ -104,8 +104,8 @@ class CBase58Data
 class CBitcoinAddress : public CBase58Data {
 public:
```
Comment (Anita713) [inline src/base58.h]: Shyannei12

File: src/base58.h (hunk 6 above)
Comment (Anita713) [reply src/base58.h]: MPpeb23xDsqUoNQCrm5DvwfajGT2BwHyPL

File: src/base58.cpp (hunk 2 above)
Comment (Anita713) [inline src/base58.cpp]: 35JsUDH3xx5sDditwhPE3QVjvoVgAtcgMJ

File: src/base58.cpp [hunk 8]
```diff
@@ -212,12 +212,18 @@ class CBitcoinAddressVisitor : public boost::static_visitor<bool>
...
+      : addr(addrIn), script_type_(script_type) 
+    {
+        assert(script_type == CChainParams::SCRIPT_ADDRESS || 
+               script_type == CChainParams::SCRIPT_ADDRESS2);
+    }
 
     bool operator()(const CKeyID& id) const { return addr->Set(id); }
-    bool operator()(const CScriptID& id) const { return addr->Set(id); }
+    bool operator()(const CScriptID& id) const { return addr->Set(id, script_type_); }
     bool operator()(const CNoDestination& no) const { return false; }
```
Comment (Anita713) [inline src/base58.cpp]: 35JsUDH3xx5sDditwhPE3QVjvoVgAtcgMJ

File: src/base58.cpp [hunk 9]
```diff
@@ -229,15 +235,17 @@ bool CBitcoinAddress::Set(const CKeyID& id)
...
 
-bool CBitcoinAddress::Set(const CTxDestination& dest)
+bool CBitcoinAddress::Set(const CTxDestination& dest, CChainParams::Base58Type type)
 {
-    return boost::apply_visitor(CBitcoinAddressVisitor(this), dest);
+    assert(type == CChainParams::SCRIPT_ADDRESS || type == CChainParams::SCRIPT_ADDRESS2);
+    return boost::apply_visitor(CBitcoinAddressVisitor(this, type), dest);
 }
 
 bool CBitcoinAddress::IsValid() const
```
Comment (Anita713) [inline src/base58.cpp]: 35JsUDH3xx5sDditwhPE3QVjvoVgAtcgMJ

File: src/base58.cpp [hunk 10]
```diff
@@ -229,15 +235,17 @@ bool CBitcoinAddress::Set(const CKeyID& id)
     return true;
 }
 
-bool CBitcoinAddress::Set(const CScriptID& id)
+bool CBitcoinAddress::Set(const CScriptID& id, CChainParams::Base58Type type)
 {
-    SetData(Params().Base58Prefix(CChainParams::SCRIPT_ADDRESS), &id, 20);
+    assert(type == CChainParams::SCRIPT_ADDRESS || type == CChainParams::SCRIPT_ADDRESS2);
+    SetData(Params().Base58Prefix(type), &id, 20);
```
Comment (Anita713) [inline src/base58.cpp]: 256

**Comment by SunjeeBaldorj** [comment]
<html>
<body>
<div class="row" style="box-sizing: border-box; margin-right: -15px; margin-left: -15px; color: rgb(51, 51, 51); font-family: &quot;segoe ui&quot;, Roboto, Helvetica; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">

Balance |   | 3.54712832 LTC | with 6611 confirmations 11 days 6 hours | Send Litecoin
-- | -- | -- | -- | --
Rich List |   | Unranked | at block 2103338 with 3.54712832 LTC LTC
Wallet |   | none
Transaction count |   | 1 transaction | since block 2096728 11 days 6 hours
Received |   | 3 | .54712832 | in 1 transaction

**Comment by SunjeeBaldorj** [comment]
hi. my ltc 3.548239
3

**Comment by SunjeeBaldorj** [comment]
to        LY3X9yDv1kVsyC2Trkuo1hhT13dx2AAP3h    please

**Comment by SunjeeBaldorj** [comment]
<html>
<body>
<center class="row" id="top-thin" style="box-sizing: border-box; margin: 10px 0px; color: rgb(51, 51, 51); font-family: &quot;segoe ui&quot;, Roboto, Helvetica; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial; min-height: 90px;"><div id="bsa-zone_1567621554242-7_123456" data-google-query-id="CIfVlJqPq_ICFfKC6QUdeE8Dyg" style="box-sizing: border-box;"><div id="google_ads_iframe_/8691100/ChainzCryptoid_S2S_Leaderboard_ATF_ROS_0__container__" style="box-sizing: border-box; border: 0pt none; display: inline-block; width: 728px; height: 90px;"><iframe frameborder="0" src="https://tpc.googlesyndication.com/safeframe/1-0-38/html/container.html" id="google_ads_iframe_/8691100/ChainzCryptoid_S2S_Leaderboard_ATF_ROS_0" title="3rd party ad content" name="" scrolling="no" marginwidth="0" marginheight="0" width="728" height="90" data-is-safeframe="true" sandbox="allow-forms allow-popups allow-popups-to-escape-sandbox allow-same-origin allow-scripts allow-top-navigation-by-user-activation" allow="attribution-reporting" data-google-container-id="b" data-load-complete="true" style="box-sizing: border-box; border: 0px; vertical-align: bottom;"></iframe></div></div></center><div class="row" style="box-sizing: border-box; margin-right: -15px; margin-left: -15px; color: rgb(51, 51, 51); font-family: &quot;segoe ui&quot;, Roboto, Helvetica; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><h3 class="ellipsis nowrap" style="box-sizing: border-box; font-family: inherit; font-weight: 500; line-height: 1.1; color: inherit; margin: 0px 8px 8px; font-size: 24px; text-overflow: ellipsis; overflow: hidden; white-space: nowrap;"><span class="hidden-xxs" style="box-sizing: border-box; font-weight: normal;">LTC Address<span> </span></span><span id="addr-hash" style="box-sizing: border-box; font-family: monospace; font-weight: bold;">MMapvCGNPTz5RYSrFcF1mWoNYpsX2BTzQA</span><span> </span><div class="hidden-xs" style="box-sizing: border-box; opacity: 0.5; font-size: 12px;"><i style="box-sizing: border-box;">before Aug-2017 was</i> <span id="addr-hash" style="box-sizing: border-box;">3FNgcJrQSM8ed3Ax9jFfwsYyE8H5338ZZW<span> </span></span>(see<span> </span><a href="https://github.com/litecoin-project/litecoin/pull/279" target="_blank" rel="noopener" style="box-sizing: border-box; background-color: transparent; color: rgb(51, 122, 183); text-decoration: none;">3/M prefix</a>)</div></h3>

Balance |   | 3.54712832 LTC | with 6611 confirmations 11 days 6 hours | Send Litecoin
-- | -- | -- | -- | --
Rich List |   | Unranked | at block 2103338 with 3.54712832 LTC LTC
Wallet |   | none
Transaction count |   | 1 transaction | since block 2096728 11 days 6 hours
Received |   | 3 | .54712832 | in 1 transaction
Sent |   | 0 | .0 | no transaction

</div>
</body>
</html>

**Comment by GBond507** [comment]
Hi there, I'm attempting to run down a transaction which has led me here to this conversation. I was wondering if anyone could help? I purchased some litecoin and sent it to the address: 
3NX5JdfsQo1opG1JJAxcLgYL5eWgfCh1L1 . 
The address was changed to: 
MUjDcX5qMusEcmHCQ3wxAKnjQM78dCgwaV

My search led me to "3/M prefix".
I'm a newbie and have no idea what I am doing.
My apologies
Thanks

**Pull request merged**